examples/python/Boxes_QPE_example.py

```diff
@@ -0,0 +1,561 @@
+    qft_box = CircBox(build_qft_circuit(n_measurement_qubits))
+    inverse_qft_box = qft_box.dagger
+    
+    qpe_circ.add_circbox(inverse_qft_box, measurement_qubits)
+    
+    qpe_circ.measure_register(measurement_register, "c")
+    
+    return qpe_circ
+
+
+# ## Phase Estimation with a Trivial Eigenstate
+# 
+# Lets test our circuit construction by preparing a trivial $|1\rangle $eigenstate of the $\text{U1}$ gate. We can then see if our phase estimation circuit returns the expected eigenvalue.
+
+# $$
+# \begin{equation}
+# U1(\phi)|1\rangle = e^{i\phi} = e^{2 \pi i \theta} \implies \theta = \frac{\phi}{2}
+# \end{equation}
+# $$
+# 
+# So we expect that our ideal phase $\theta$ will be half the input angle $\phi$ to our $U1$ gate.
+
+# In[15]:
+
+
+state_prep_circuit = Circuit(1).X(0)
+
+input_angle = 0.73
+
+unitary_circuit = Circuit(1).add_gate(OpType.U1, [input_angle], [0])
+
+
+# In[16]:
+
+
+qpe_circ_trivial = build_phase_est_circuit(4, state_prep_circuit=state_prep_circuit, unitary_circuit=unitary_circuit)
+
+
+# In[17]:
+
+
+render_circuit_jupyter(qpe_circ_trivial)
+
+
+# Lets use the noiseless `AerBackend` simulator to run our phase estimation circuit.
+
+# In[18]:
+
+
+from pytket.extensions.qiskit import AerBackend
+
+backend = AerBackend()
+
+compiled_circ = backend.get_compiled_circuit(qpe_circ_trivial)
+
+
+n_shots = 1000
+result = backend.run_circuit(compiled_circ, n_shots)
+
+print(result.get_counts())
+
+
+# In[19]:
+
+
+from plotting import plot_qpe_results
+
+plot_qpe_results(result, y_limit=1.2*n_shots)
+
+
+# As expected we see one outcome with high probability. Lets now extract our approximation of $\theta$ from our output bitstrings.
+# 
+# suppose the $j$ is an integer representation of our most commonly measured bitstring.
+
+# $$
+# \begin{equation}
+# \theta_{estimate} = \frac{j}{N}
+# \end{equation}
+# $$
+
+# Here $N = 2 ^n$ where $n$ is the number of measurement qubits.
+
+# In[20]:
+
+
+from pytket.backends.backendresult import BackendResult
+
+def single_phase_from_backendresult(result: BackendResult) -> float:
+    
+    # Extract most common measurement outcome
+    basis_state = result.get_counts().most_common()[0][0] 
+    bitstring = "".join([str(bit) for bit in basis_state])
+    
+    # Calculate theta
+    n_measurement_qubits = len(bitstring)
+    phase_estimate = int(bitstring, 2) / (2**n_measurement_qubits)
+    
+    return phase_estimate
+
+
+# In[21]:
+
+
+theta = single_phase_from_backendresult(result)
+
+
+# In[22]:
+
+
+print(theta)
+
+
+# Our output is close to half our input angle $\phi$ as expected. Lets calculate our error.
+
+# In[23]:
+
+
+error = round(abs(input_angle - (2*theta)), 5) # 
+print(error)
+
+
+# ## State Preparation
+
+# In[24]:
+
+
+import numpy as np
+import scipy
+
+
+# In[25]:
+
+
+h2_matrix = qpo_h2.to_sparse_matrix().todense()
+
+
+# In[26]:
+
+
+eigenvalues, eigenvectors = scipy.sparse.linalg.eigs(h2_matrix, k=15)
+
+
+# In[27]:
+
+
+eigenvectors[0].shape
```

Comment:
A lot of these things could be done in the same python cell? Rather than having 7 or so sequential cells?